### PR TITLE
feat: validate certfiles and keyfiles

### PR DESCRIPTION
**Pull request by thalesmg**
Fixes https://emqx.atlassian.net/browse/EMQX-10210

Release version: v/e5.9.0

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

**Review by zmstone — COMMENTED** [review]
it's hard to determine to what extent should we validate in EMQX schema checks.
but since we are decoding the files, why not call public_key:pkix_verify/2 to verify if the cert and key are truly a pair.

there are a few other things can be checked, but I'm fine with not going too far, but maybe worth mentioning in the documents:
- Certificate expiry date (since it's more for the peer to decide if expired cert is acceptable or not).
- Integrity between certificate and configured cipher suites.

File: apps/emqx_ft/test/emqx_ft_test_helpers.erl [hunk 1]
```diff
@@ -157,6 +157,29 @@ to_bin(Val) ->
 aws_config() ->
     emqx_s3_test_helpers:aws_config(tcp, binary_to_list(?S3_HOST), ?S3_PORT).
 
+pem_certfile() ->
```
Comment (zmstone) [inline apps/emqx_ft/test/emqx_ft_test_helpers.erl:160]: maybe call emqx_test_tls_certs_helper to generate test certificate.

File: apps/emqx/test/emqx_tls_lib_tests.erl [hunk 2]
```diff
@@ -470,3 +676,69 @@ test_name_hash_key() ->
         "RFY7JjluKcVkp/zCDeUxTU3O6sS+v6/3VE11Cob6OYQx3lN5wrZ3\n"
         "-----END RSA PRIVATE KEY-----\n"
     >>.
+
+setup_ssl_files() ->
+    Dir = mktemp_dir(),
+    DefaultPassword = <<"foobar">>,
+    emqx_test_tls_certs_helper:gen_ca(Dir, "root"),
```
Comment (zmstone) [inline apps/emqx/test/emqx_tls_lib_tests.erl:683]: emqx_test_tls_certs_helper only implemented ec keys, maybe test RSA keys too ?

File: apps/emqx/src/emqx_tls_lib.erl [hunk 3]
```diff
@@ -420,6 +432,51 @@ is_pem(MaybePem) ->
...
+try_validate_pem(PEM, certfile, _Password) ->
+    do_validate_certfile(PEM);
+try_validate_pem(PEM, keyfile, Password) ->
+    do_validate_keyfile(PEM, Password);
+try_validate_pem(_PEM, _Type, _Password) ->
+    ok.
+
+do_validate_certfile(PEM) ->
+    maybe
+        {ok, [{'Certificate' = Type, DER, not_encrypted}]} ?= ?catching(public_key:pem_decode(PEM)),
```
[the remark this answers is not in the document]
Comment (thalesmg) [reply apps/emqx/src/emqx_tls_lib.erl:452]: does that apply to certificates as well?  or only CA certificates?

**Comment by thalesmg** [comment]
> why not call public_key:pkix_verify/2 to verify if the cert and key are truly a pair.

that's slightly trickier: it requires obtaining the parsed public key (not just the der binary) for different types of keys (rsa, ec, whatever), and OTP doesn't seem to provide an unified way to do that.  I'll see if I can find a way...

File: apps/emqx/src/emqx_tls_lib.erl (hunk 3 above)
[the remark this answers is not in the document]
Comment (zmstone) [reply apps/emqx/src/emqx_tls_lib.erl:452]: yes, certificate file should be the chain, cacertfile is a collection of root CA certificates, not a chain.

**Comment by thalesmg** [comment]
> > why not call public_key:pkix_verify/2 to verify if the cert and key are truly a pair.
> 
> that's slightly trickier: it requires obtaining the parsed public key (not just the der binary) for different types of keys (rsa, ec, whatever), and OTP doesn't seem to provide an unified way to do that. I'll see if I can find a way...

actually, `pkix_verify/2` only applies to RSA keys, according to the typespec, so it's not general for EC keys.

https://github.com/erlang/otp/blob/9ae2ef54fc6dacb5bfbe2dc6fc6f2522d57ce1fb/lib/public_key/src/public_key.erl#L908

**Review by zmstone — COMMENTED** [review]
It seems there is no longer the need of the `is_pem/1` implementation?
if try_validate_pem_file will try to decode it anyway.

**Comment by thalesmg** [comment]
> It seems there is no longer the need of the `is_pem/1` implementation? if try_validate_pem_file will try to decode it anyway.

I was avoiding to remove this because some places seem to expect this `pem_check` key, and also it serves to differentiate between a filepath and the pem contents.  :thinking:

**Review by zmstone — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-6665)

**Pull request merged**